src/bert_record_validator.erl

```diff
@@ -0,0 +1,216 @@
+-module(bert_record_validator).
+-include("bert.hrl").
+%%  - Extend the source code with a validate/1 and validate/2 function.
+%%  - Add a function that return all records that are validated with this program
+-export([parse_transform/2]).
+-include_lib("syntax_tools/include/merl.hrl").
+-define(QUOTE(X), ?Q(??X)).
+-compile(export_all).
+
+to_upper(Name) when is_atom(Name) ->
+    to_upper(atom_to_list(Name));
+to_upper(Name) ->
+    [Hd | Tail] = Name,
+    [string:to_upper(Hd)] ++ Tail.
+
+%% Add vlidation functions for header files included in source
+parse_transform(Forms, Options) ->
+    AF = erl_syntax_lib:analyze_forms(Forms),
+    Module = proplists:get_value(module, AF),
+    CompileOpts =
+        lists:foldl(fun({compile, Opts}, Acc) when is_list(Opts) -> Acc ++ Opts;
+                       ({compile, Opt}, Acc) -> Acc ++ [Opt];
+                       (_, Acc) -> Acc
+                    end, Options, proplists:get_value(attributes, AF, [])),
+
+    %% We keep a copy of the generated source code, but that's actually not needed
+    Dest = proplists:get_value(bert_erl, CompileOpts, ?ERL),
+    File = filename:join(Dest, lists:concat([Module, ".erl"])),
+    filelib:ensure_dir(File),
+
+    {Records, Validate} = directives(Forms),
+    ExtraForms =
+        erl_syntax:revert_forms(
+          [?QUOTE(validate(Obj) ->
+                         case invalid_fields(Obj, []) of
+                             [] -> [];
+                             Errors -> {error, Errors}
+                         end.),
+           ?QUOTE(validatable() -> _@Records@.),
+           Validate]),
+    NewForms = lists:keydelete(eof, 1, Forms) ++ [?QUOTE(-export([validate/1, validatable/0]).)] ++ ExtraForms,
+    file:write_file(File, ["%% Code generated !\n", erl_prettypr:format(erl_syntax:form_list(NewForms))]),
```

Comment:
`Generated code (do NOT edit)` or something is the normal thing to say?

Reply:
In the slightly longer run,I would really like toremove the generated file. Generated source code will eventually lead to trouble. Ileft it in for themoment such that we can compare and inspect.

I the long run, you would use the compile option 'P' to find the source code.